### PR TITLE
Supporting helm3 to show up resource names that were deployed as part…

### PR DESCRIPTION
**Pull request by sarvaniy**
… of release in helm status command

### CODE REVIEW

**Comment by bacongobbler** [comment]
Hi @sarvaniy. Would you mind looking into the failing tests?

**Review by bacongobbler — CHANGES_REQUESTED** [review]

**Comment by bacongobbler** [comment]
I started playing around with this to see what we can do.

```
import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"strings"
	"sync"
	"time"

	jsonpatch "github.com/evanphx/json-patch"
	"github.com/pkg/errors"
	batch "k8s.io/api/batch/v1"
	v1 "k8s.io/api/core/v1"
	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
	apiextv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/api/meta"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/fields"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/apimachinery/pkg/util/strategicpatch"
	"k8s.io/apimachinery/pkg/watch"
	"k8s.io/cli-runtime/pkg/genericclioptions"
	genericprinters "k8s.io/cli-runtime/pkg/printers"
	"k8s.io/cli-runtime/pkg/resource"
	"k8s.io/client-go/kubernetes/scheme"
	cachetools "k8s.io/client-go/tools/cache"
	watchtools "k8s.io/client-go/tools/watch"
	cmdutil "k8s.io/kubectl/pkg/cmd/util"
	"k8s.io/kubernetes/pkg/printers"
	"k8s.io/kubernetes/pkg/printers/internalversion"
)

...

func (c *Client) Get(resources ResourceList) (string, error) {
	builder := strings.Builder{}
	printer := genericprinters.NewTablePrinter(genericprinters.PrintOptions{})
	generator := printers.NewTablePrinter()
	internalversion.AddHandlers(generator)
	err := resources.Visit(func(info *resource.Info, err error) error {
		if err != nil {
			return err
		}

		obj := AsVersioned(info)
		if err != nil {
			if !apierrors.IsNotFound(err) {
				return errors.Wrap(err, "could not get information about the resource")
			}
			return err
		}
		tbl, err := generator.PrintTable(obj, printers.PrintOptions{})
		if err != nil {
			return err
		}
		if err := printer.PrintObj(tbl, &builder); err != nil {
			return fmt.Errorf("failed to print object %s: %v", info.Name, err)
		}
		if _, err := builder.WriteString("\n"); err != nil {
			return err
		}
		return nil
	})
	return builder.String(), err
}
```

In order to get this to work, this code would need to rely on packages internal to kubernetes/kubernetes, in particular the `TableGenerator` as part of `k8s.io/kubernetes/pkg/printers`: https://github.com/kubernetes/kubernetes/blob/master/pkg/printers/tablegenerator.go

I'm not sure this can reasonably be achieved without reintroducing the entire kubernetes client library code, which we went to great lengths to remove in Helm 3.

For now, `helm get manifest | kubectl get -f -` works just fine.

Let us know if you find out something different.

**Comment by sarvaniy** [comment]
@bacongobbler Thanks for update, as you suggest I will try to look into kubectl get implementation in deeper.

**Comment by sarvaniy** [comment]
@bacongobbler Hi, I had gone through "kubectl get" implementation in depper. I found that Transformrequests to get the status of resources instead of rely on packages internal to kubernetes/kubernetes. 

I had modified the code and tested, it worked fine.

example:
helm status prometheus -n monitoring

NAME: prometheus
LAST DEPLOYED: Wed Dec 11 15:59:46 2019
NAMESPACE: monitoring
STATUS: deployed
REVISION: 1
RESOURCES:
==> v1/ServiceAccount
NAME                      SECRETS   AGE
prometheus-alertmanager   1         88d
prometheus-kube-state-metrics   1     88d
prometheus-node-exporter   1     88d
prometheus-pushgateway   1     88d
prometheus-server   1     88d

==> v1beta1/ClusterRole
NAME                      AGE
prometheus-alertmanager   88d
prometheus-kube-state-metrics   88d
prometheus-pushgateway   88d
prometheus-server   88d

==> v1beta1/ClusterRoleBinding
prometheus-alertmanager   88d
prometheus-kube-state-metrics   88d
prometheus-pushgateway   88d
prometheus-server   88d

==> v1/Service
NAME                      TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE
prometheus-alertmanager   ClusterIP   10.109.247.218   <none>        80/TCP    88d
prometheus-kube-state-metrics   ClusterIP   None   <none>   80/TCP   88d
prometheus-node-exporter   ClusterIP   None   <none>   9100/TCP   88d
prometheus-pushgateway   ClusterIP   10.101.68.104   <none>   9091/TCP   88d
prometheus-server   ClusterIP   10.99.38.220   <none>   80/TCP   88d

==> v1/DaemonSet
NAME                       DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
prometheus-node-exporter   8         8         8       8            8           <none>          88d

==> v1/Deployment
NAME                      READY   UP-TO-DATE   AVAILABLE   AGE
prometheus-alertmanager   1/1     1            1           88d
prometheus-kube-state-metrics   1/1   1     1     88d
prometheus-pushgateway   1/1   1     1     88d
prometheus-server   1/1   1     1     88d

==> v1/ConfigMap
NAME                      DATA   AGE
prometheus-alertmanager   1      88d
prometheus-server   5     88d

==> v1/PersistentVolumeClaim
NAME                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
prometheus-alertmanager   Bound    pvc-29b9c900-1c59-11ea-8152-0cc47af2ef34   2Gi        RWO            basic          88d
prometheus-server   Bound   pvc-29b9c4c1-1c59-11ea-8152-0cc47af2ef34   8Gi   RWO   basic   88d


TEST SUITE: None
NOTES:
The Prometheus server can be accessed via port 80 on the following DNS name from within your cluster:
prometheus-server.monitoring.svc.cluster.local


Get the Prometheus server URL by running these commands in the same shell:
  export POD_NAME=$(kubectl get pods --namespace monitoring -l "app=prometheus,component=server" -o jsonpath="{.items[0].metadata.name}")
  kubectl --namespace monitoring port-forward $POD_NAME 9090


The Prometheus alertmanager can be accessed via port 80 on the following DNS name from within your cluster:
prometheus-alertmanager.monitoring.svc.cluster.local


Get the Alertmanager URL by running these commands in the same shell:
  export POD_NAME=$(kubectl get pods --namespace monitoring -l "app=prometheus,component=alertmanager" -o jsonpath="{.items[0].metadata.name}")
  kubectl --namespace monitoring port-forward $POD_NAME 9093
#################################################################################
######   WARNING: Pod Security Policy has been moved to a global property.  #####
######            use .Values.podSecurityPolicy.enabled with pod-based      #####
######            annotations                                               #####
######            (e.g. .Values.nodeExporter.podSecurityPolicy.annotations) #####
#################################################################################


The Prometheus PushGateway can be accessed via port 9091 on the following DNS name from within your cluster:
prometheus-pushgateway.monitoring.svc.cluster.local


Get the PushGateway URL by running these commands in the same shell:
  export POD_NAME=$(kubectl get pods --namespace monitoring -l "app=prometheus,component=pushgateway" -o jsonpath="{.items[0].metadata.name}")
  kubectl --namespace monitoring port-forward $POD_NAME 9091

For more information on running Prometheus, visit:
https://prometheus.io/




And I had Commited the same tested code.

File: pkg/kube/client.go
Comment (mattfarina) [inline pkg/kube/client.go]: If this is used with the normal display this would be fine. But, if JSON or YAML is the output the complex string is going to end up inside a key and not be easily parsed.

I would suggest getting the output should be different per type and a print time operation depending on the output format. This would be more useful for getting this information in a machine format.

**Review by mattfarina — CHANGES_REQUESTED** [review]
I really like that this is being worked on. I know of tools that use this and parse the information.

But, to make it better with JSON/YAML output I think there can be a better option... see below...

File: pkg/kube/client.go
Comment (sarvaniy) [reply pkg/kube/client.go]: Hi Matt,

Commited code is worked fine for different output formats like yaml and json as well.

Below is the genearted output without format and with different format cases

**Without output format case**

helm status firsthelm-0-1579514513 
```
NAME: firsthelm-0-1579514513
LAST DEPLOYED: Mon Jan 20 05:01:57 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
RESOURCES:
==> v1/Service
NAME                     TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
firsthelm-0-1579514513   ClusterIP   10.96.41.237   <none>        80/TCP    50d

==> v1/Deployment
NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
firsthelm-0-1579514513   1/1     1            1           50d


NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=firsthelm,app.kubernetes.io/instance=firsthelm-0-1579514513" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:80
```

**With yaml as output format**

helm status firsthelm-0-1579514513 **-o yaml**

```
Resources: |2+

  ==> v1/Service
  NAME                     TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
  firsthelm-0-1579514513   ClusterIP   10.96.41.237   <none>        80/TCP    50d

  ==> v1/Deployment
  NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
  firsthelm-0-1579514513   1/1     1            1           50d

hooks:
- events:
  - test
  kind: Pod
  last_run:
    completed_at: ""
    phase: ""
    started_at: ""
  manifest: |-
    apiVersion: v1
    kind: Pod
    metadata:
      name: "firsthelm-0-1579514513-test-connection"
      labels:
        app.kubernetes.io/name: firsthelm
        helm.sh/chart: firsthelm-0.1.0
        app.kubernetes.io/instance: firsthelm-0-1579514513
        app.kubernetes.io/version: "1.0"
        app.kubernetes.io/managed-by: Helm
      annotations:
        "helm.sh/hook": test-success
    spec:
      containers:
        - name: wget
          image: busybox
          command: ['wget']
          args:  ['firsthelm-0-1579514513:80']
      restartPolicy: Never
  name: firsthelm-0-1579514513-test-connection
  path: firsthelm/templates/tests/test-connection.yaml
info:
  deleted: ""
  description: Install complete
  first_deployed: "2020-01-20T05:01:57.370835613-05:00"
  last_deployed: "2020-01-20T05:01:57.370835613-05:00"
  notes: |
    1. Get the application URL by running these commands:
      export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=firsthelm,app.kubernetes.io/instance=firsthelm-0-1579514513" -o jsonpath="{.items[0].metadata.name}")
      echo "Visit http://127.0.0.1:8080 to use your application"
      kubectl port-forward $POD_NAME 8080:80
  status: deployed
manifest: |
  ---
  # Source: firsthelm/templates/service.yaml
  apiVersion: v1
  kind: Service
  metadata:
    name: firsthelm-0-1579514513
    labels:
      app.kubernetes.io/name: firsthelm
      helm.sh/chart: firsthelm-0.1.0
      app.kubernetes.io/instance: firsthelm-0-1579514513
      app.kubernetes.io/version: "1.0"
      app.kubernetes.io/managed-by: Helm
  spec:
    type: ClusterIP
    ports:
      - port: 80
        targetPort: http
        protocol: TCP
        name: http
    selector:
      app.kubernetes.io/name: firsthelm
      app.kubernetes.io/instance: firsthelm-0-1579514513
  ---
  # Source: firsthelm/templates/deployment.yaml
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: firsthelm-0-1579514513
    labels:
      app.kubernetes.io/name: firsthelm
      helm.sh/chart: firsthelm-0.1.0
      app.kubernetes.io/instance: firsthelm-0-1579514513
      app.kubernetes.io/version: "1.0"
      app.kubernetes.io/managed-by: Helm
  spec:
    replicas: 1
    selector:
      matchLabels:
        app.kubernetes.io/name: firsthelm
        app.kubernetes.io/instance: firsthelm-0-1579514513
    template:
      metadata:
        labels:
          app.kubernetes.io/name: firsthelm
          app.kubernetes.io/instance: firsthelm-0-1579514513
      spec:
        containers:
          - name: firsthelm
            image: "nginx:stable"
            imagePullPolicy: IfNotPresent
            ports:
              - name: http
                containerPort: 80
                protocol: TCP
            livenessProbe:
              httpGet:
                path: /
                port: http
            readinessProbe:
              httpGet:
                path: /
                port: http
            resources:
              {}
name: firsthelm-0-1579514513
namespace: default
version: 1
```

**With json as output format**

helm status firsthelm-0-1579514513  **-o json**
```
{"name":"firsthelm-0-1579514513","info":{"first_deployed":"2020-01-20T05:01:57.370835613-05:00","last_deployed":"2020-01-20T05:01:57.370835613-05:00","deleted":"","description":"Install complete","status":"deployed","notes":"1. Get the application URL by running these commands:\n  export POD_NAME=$(kubectl get pods --namespace default -l \"app.kubernetes.io/name=firsthelm,app.kubernetes.io/instance=firsthelm-0-1579514513\" -o jsonpath=\"{.items[0].metadata.name}\")\n  echo \"Visit http://127.0.0.1:8080 to use your application\"\n  kubectl port-forward $POD_NAME 8080:80\n"},"manifest":"---\n# Source: firsthelm/templates/service.yaml\napiVersion: v1\nkind: Service\nmetadata:\n  name: firsthelm-0-1579514513\n  labels:\n    app.kubernetes.io/name: firsthelm\n    helm.sh/chart: firsthelm-0.1.0\n    app.kubernetes.io/instance: firsthelm-0-1579514513\n    app.kubernetes.io/version: \"1.0\"\n    app.kubernetes.io/managed-by: Helm\nspec:\n  type: ClusterIP\n  ports:\n    - port: 80\n      targetPort: http\n      protocol: TCP\n      name: http\n  selector:\n    app.kubernetes.io/name: firsthelm\n    app.kubernetes.io/instance: firsthelm-0-1579514513\n---\n# Source: firsthelm/templates/deployment.yaml\napiVersion: apps/v1\nkind: Deployment\nmetadata:\n  name: firsthelm-0-1579514513\n  labels:\n    app.kubernetes.io/name: firsthelm\n    helm.sh/chart: firsthelm-0.1.0\n    app.kubernetes.io/instance: firsthelm-0-1579514513\n    app.kubernetes.io/version: \"1.0\"\n    app.kubernetes.io/managed-by: Helm\nspec:\n  replicas: 1\n  selector:\n    matchLabels:\n      app.kubernetes.io/name: firsthelm\n      app.kubernetes.io/instance: firsthelm-0-1579514513\n  template:\n    metadata:\n      labels:\n        app.kubernetes.io/name: firsthelm\n        app.kubernetes.io/instance: firsthelm-0-1579514513\n    spec:\n      containers:\n        - name: firsthelm\n          image: \"nginx:stable\"\n          imagePullPolicy: IfNotPresent\n          ports:\n            - name: http\n              containerPort: 80\n              protocol: TCP\n          livenessProbe:\n            httpGet:\n              path: /\n              port: http\n          readinessProbe:\n            httpGet:\n              path: /\n              port: http\n          resources:\n            {}\n","hooks":[{"name":"firsthelm-0-1579514513-test-connection","kind":"Pod","path":"firsthelm/templates/tests/test-connection.yaml","manifest":"apiVersion: v1\nkind: Pod\nmetadata:\n  name: \"firsthelm-0-1579514513-test-connection\"\n  labels:\n    app.kubernetes.io/name: firsthelm\n    helm.sh/chart: firsthelm-0.1.0\n    app.kubernetes.io/instance: firsthelm-0-1579514513\n    app.kubernetes.io/version: \"1.0\"\n    app.kubernetes.io/managed-by: Helm\n  annotations:\n    \"helm.sh/hook\": test-success\nspec:\n  containers:\n    - name: wget\n      image: busybox\n      command: ['wget']\n      args:  ['firsthelm-0-1579514513:80']\n  restartPolicy: Never","events":["test"],"last_run":{"started_at":"","completed_at":"","phase":""}}],"version":1,"namespace":"default","Resources":"\n==\u003e v1/Service\nNAME                     TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE\nfirsthelm-0-1579514513   ClusterIP   10.96.41.237   \u003cnone\u003e        80/TCP    50d\n\n==\u003e v1/Deployment\nNAME                     READY   UP-TO-DATE   AVAILABLE   AGE\nfirsthelm-0-1579514513   1/1     1            1           50d\n\n"}
```


I think that, in helm2 also: "helm status" command with different output formats, the output is same.

**Comment by sarvaniy** [comment]
@bacongobbler , @mattfarina 

Hi All,

   Could you please review the code and comments. Thanks in advance

File: pkg/kube/client.go
Comment (sarvaniy) [reply pkg/kube/client.go]: @bacongobbler 

Hi,
    I had modified the code as per your suggestion. I had gone through "kubectl get" implementation in depper. I found that Transformrequests is used to get the status of resources instead of rely on packages internal to kubernetes/kubernetes.

I had modified the code and tested, it worked fine. Could you please review it. You can check the resources output in below conservations.

**Comment by sarvaniy** [comment]
@bacongobbler , @mattfarina

Hi All,

        Could you please review the code and comments. Thanks in advance

File: pkg/kube/client.go
Comment (mattfarina) [inline pkg/kube/client.go]: This should also accept the v1 as well as the v1beta1. You can see the work I did for Helm v2 on this at #7785.

**Review by mattfarina — CHANGES_REQUESTED** [review]
Thanks for the continued work on this. I just worked out how to do this with Helm v2 (see PR link in comments). Helm v2 has a simpler situation as the JSON/YAML output we expect to be different. I commented on the previous thread about that as well.

File: pkg/kube/client.go
Comment (mattfarina) [reply pkg/kube/client.go]: The release part of this is a string of the output. Instead, we should have that as structured metadata. It doesn't need to be the same as the output of `kubectl get po --output json`. I would expect the structure to be a JSON object that represents the table as JSON.

Right now (in Helm v2) machines need to use a regex to parse the content of the status information. That includes getting information about things like pods. If this was represented as JSON/YAML when the output was JSON or YAML it would make machines using this easier.

File: pkg/kube/client.go
Comment (sarvaniy) [reply pkg/kube/client.go]: Hi Matt,

   Thanks for the explanation. As you suggest, i made changes inorder to support both metav1 and metav1beta1 cases. I had gone through the above PR link and used the structured metadata for JSON/YAML. Everything is working fine.

Below are the code-snippet changes.

```
func transformRequests(req *rest.Request) {
        tableParam := strings.Join([]string{
                fmt.Sprintf("application/json;as=Table;v=%s;g=%s", metav1.SchemeGroupVersion.Version, metav1.GroupName),
                fmt.Sprintf("application/json;as=Table;v=%s;g=%s", metav1beta1.SchemeGroupVersion.Version, metav1beta1.GroupName),
                "application/json",
        }, ",")
        req.SetHeader("Accept", tableParam)

        // if sorting, ensure we receive the full object in order to introspect its fields via jsonpath
        req.Param("includeObject", "Object")
}

func sortTableSlice(objs []runtime.Object) []runtime.Object {
        if len(objs) < 2 {
                return objs
        }

        ntbl := &metav1.Table{}
        // Sort the list of objects
        var newObjs []runtime.Object
        namesCache := make(map[string]runtime.Object, len(objs))
        var names []string
        for _, obj := range objs {
                unstr, ok := obj.(*unstructured.Unstructured)
                if !ok {
                        return objs
                }
                if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstr.Object, ntbl); err != nil {
                        return objs
                }
                namesCache[ntbl.GetSelfLink()] = obj
                names = append(names, ntbl.GetSelfLink())
        }
        sort.Strings(names)
        for _, name := range names {
                newObjs = append(newObjs, namesCache[name])
        }
        return newObjs
}


func (c *Client) Get(resources ResourceList) (string, error) {
        buf := new(bytes.Buffer)
        printFlags := get.NewHumanPrintFlags()
        typePrinter, _ := printFlags.ToPrinter("")
        printer := &get.TablePrinter{Delegate: typePrinter}
        objs := make(map[string][]runtime.Object)
        err := resources.Visit(func(info *resource.Info, err error) error {
                if err != nil {
                        return err
                }
                gvk := info.ResourceMapping().GroupVersionKind
                vk := gvk.Version + "/" + gvk.Kind
                obj, err := getResource(info)
                if err != nil {
                        return err
                }
                objs[vk] = append(objs[vk], obj)
                return nil
        })
        if err != nil {
                return "", err
        }
        var sortingkeys []string
        for key := range objs {
                sortingkeys = append(sortingkeys, key)
        }
        // Sort alphabetically by version/kind keys
        sort.Strings(sortingkeys)
        for _, t := range sortingkeys {
                if _, err = fmt.Fprintf(buf, "==> %s\n", t); err != nil {
                        return "", err
                }
                vk := objs[t]
                vk = sortTableSlice(vk)
                for _, resource := range vk {
                        if err := printer.PrintObj(resource, buf); err != nil {
                                c.Log("failed to print object type %s: %v", t, err)
                                return "", err
                        }
                }
                if _, err := buf.WriteString("\n"); err != nil {
                        return "", err
                }
        }
        return buf.String(), nil
}

func getResource(info *resource.Info) (runtime.Object, error) {
        obj, err := resource.NewHelper(info.Client, info.Mapping).Get(info.Namespace, info.Name, false)
        if err != nil {
                return nil, err
        }
        return obj, nil
}
```

Could you please look into it and let me know is this okay or not. If yes, then i will commit the code changes.

**Comment by sarvaniy** [comment]
Hi Matt, 
    Could you please look into the PR comments, based on that I will commit my code. Thanks in advance

File: pkg/kube/client.go
Comment (mattfarina) [reply pkg/kube/client.go]: This part looks ok. Once I have the PR to review in a full code editor I can do a better review. I look forward to seeing your changes.

**Comment by sarvaniy** [comment]
Thanks Matt

**Comment by sarvaniy** [comment]
I will do Matt. :)

**Comment by sarvaniy** [comment]
Hi Matt,

    I had commited the changes. Could you please review the code.

Thanks.

**Comment by sarvaniy** [comment]
Hi Matt,

I had commited the changes. Could you please review the code.

Thanks.

File: pkg/kube/client.go
Comment (sarvaniy) [reply pkg/kube/client.go]: Hi Matt,

   I had commited the changes. Could you please review the code.

**Comment by svolence** [comment]
> Hi Matt,
> 
> I had commited the changes. Could you please review the code.
> 
> Thanks.

merged now ?

**Comment by sarvaniy** [comment]
Hi svolence,

   Just now , i had resolved the merge conflicts, you can merge it. Thanks for the response.

**Comment by sarvaniy** [comment]
> > Hi Matt,
> > I had commited the changes. Could you please review the code.
> > Thanks.
> 
> merged now ?

yes svolence, you can merge it. Thanks for the response.

**Comment by r3dlight** [comment]
Hi,
If this MR is ok for you guys, could you merge it  please ? 
I'm also waiting for it :) thx

**Comment by sarvaniy** [comment]
Hi,

 As no one is responding to this,  if everything is fine Could you please merge it.

**Comment by bacongobbler** [comment]
A couple of things here are preventing us from merging this:

- we cannot merge this as there is a merge conflict with the master branch.
- there is a lack of unit tests here. If we are to consider adding this functionality back, we need to have safety guarantees that version updates will not cause regressions again, which was the reason why this code was removed in the first place.

Once you address those issues we can consider merging this. Thanks.

**Comment by sarvaniy** [comment]
Hi bacongobbler,

    I had resolved the merge conflicts and code rebase is done. Currently, PR has no conflicts with base branch. Actaullly i don't know about unit tests, but based upon the provided information if unit tests are not considered, then i think so, this PR is ready to merge.

Thanks.

**Review by bacongobbler — CHANGES_REQUESTED** [review]
When this issue was first discussed, we [made it clear early on](https://github.com/helm/helm/issues/5952#issuecomment-580399492) (in fact we made that clear [multiple times](https://github.com/helm/helm/issues/5952#issuecomment-593504926)) that we would not accept PRs without some form of idea that it would not break backwards compatiblity. We need to have unit tests to ensure that is the case going forward. Without unit tests, it puts a lot of strain on the maintainers to constantly test that this code will not break between Kubernetes version bumps. The goal was to write code that doesn't happen to break every release cycle, and in its current state, this PR fails to ensure that this code will not break.

Without unit tests, we will not merge this, as we will not have a way to determine if a version bump to Kubernetes 1.19 will break this code or not.

I've already [wasted](https://github.com/helm/helm/pull/7728#issuecomment-595467975) a [lot of time](https://github.com/helm/helm/pull/7728#issuecomment-595519152) reviewing and trying to test this code manually, so having something in place to ensure this works as intended would help us out a lot.

Please add unit tests to ensure that this code will continue to work in the future. Thank you.

**Comment by sarvaniy** [comment]
Hi bacongobbler,

   Thanks for the information. I will go through how to add unit tests and i will add them for this PR.

Thank you.

**Comment by bacongobbler** [comment]
checking in to see what's the status on this PR.

**Comment by sc-chad** [comment]
The one feature I'm seriously waiting for!  @sarvaniy please let us know where you are at with this and if you need help finishing it up.  Definitely something the community is wanting.

**Comment by sarvaniy** [comment]
Hi bacongobbler,
     
      It's a bit delayed from my side and finally i added unit tests for helm status with resources(validated for normal commandline output and also with json format) and resolved merge conflicts as well. If everything is fine, then could you please merge this commit.

Best Regards,
Sarvani.

**Comment by agoloAbbas** [comment]
Thanks @sarvaniy 

I think you might need to re-request review?

**Comment by sarvaniy** [comment]
Hi @agoloAbbas,

    Thanks for suggestion, I had raised Re-request review to @bacongobbler

**Comment by bacongobbler** [comment]
Thanks. There are a little over one hundred PRs in the queue right now. Not sure when I’ll have time to review this again.

**Comment by dbsanfte** [comment]
I'm going to stick my neck out and ping @bacongobbler on this, since I know a lot of us are waiting for this feature. :)

**Comment by schollii** [comment]
@bacongobbler are you the only reviewer? A 100 PRs seems more than any one individual could handle with full-time work etc.

**Comment by mshakhmaykin** [comment]
So.. New 3.5.0 release doesn't include this. Are we now waiting till May??

**Comment by bacongobbler** [comment]
> Are we now waiting till May??

That is correct. Helm 3.6.0 is scheduled for May.

> @bacongobbler are you the only reviewer for PRs?

No. There are several Helm core maintainers, all of whom can review code. https://github.com/helm/helm/blob/master/OWNERS

Given the size of this PR, [as per policy](https://github.com/helm/helm/blob/master/CONTRIBUTING.md#size-labels) it will require two maintainers to review and sign off before this can be merged. So you'll need to find at least one other maintainer willing to review this code.

I've reviewed this PR at least 5 times over 6 months, each time either [failing to address the original issue](https://github.com/helm/helm/pull/7728#issuecomment-595467975) or [failing to provide tests](https://github.com/helm/helm/pull/7728#pullrequestreview-444144380). I outlined these issues [very early on](https://github.com/helm/helm/issues/6896#issuecomment-595519870) during the triage process, so it's suffice to say that I am VERY frustrated and burnt out from reviewing this PR.

I'd highly recommend finding two other maintainers willing to review this code.

**Comment by schollii** [comment]
Thanks for the info, I totally understand your frustration and exasperation. I will check how to flag reviewers.

File: pkg/kube/client.go
Comment (schollii) [reply pkg/kube/client.go]: @sarvaniy this seems to require fixing, are you able to do that?

**Comment by sc-chad** [comment]
Hopefully this PR doesn't die.  This is something I and others would still love to have.  Very much missed from helm2.

**Comment by koulq** [comment]
@sarvaniy hi，pod display repeats if statefulset resources are available.
image

**Comment by dbsanfte** [comment]
This feature in Helm 2 enables a lot of cool workflows like ensuring that all Helm-installed chart objects are still present in the cluster. This has saved us a few times after devs accidentally deleted a deployment :D

Helm 3 really, really needs this.

**Comment by bacongobbler** [comment]
Throwing this onto the 3.6.0 milestone for (re-)review.

**Comment by technosophos** [comment]
This adds a method to an Interface, which is defined as a breaking change by the Go authors. To refactor that, we'd have to add another interface that just has the `Get` method on it. See the Go documentation:

https://blog.golang.org/module-compatibility#TOC_3.

**Comment by mattfarina** [comment]
Unfortunately, this can't be merged. The `go.sum` file is out of date. It used to lead to dirty builds that are not caught by CI. As of Go 1.16 is leads to failures. If we merge this the tip of main will fail tests. `go mod tidy` will fix this. Moving to the next patch release.

**Comment by koulq** [comment]
This function will affect helm rollback. Executing helm rollback will delete the resource and cannot be recovered

image

**Comment by bacongobbler** [comment]
@koulq can you look through this PR and identify what might be causing that issue? How does it affect `helm rollback`?

**Comment by sc-chad** [comment]
@koulq any help here on identifying this or what's causing it?  We would love to see this get over the finish line.

**Comment by bacongobbler** [comment]
Unfortunately, it looks like this PR has become stale since the last time we provided feedback. As a result we won't be able to merge it into Helm 3.7 in its current state.

So, if someone wants to take over the work here and re-submit a new pull request with the changes suggested earlier, there is a slim chance we could bring this into Helm 3.7.

Removing from Helm 3.7's milestone for now.

**Pull request closed without merging**
